### PR TITLE
Micro app filtering feature added 

### PR DESCRIPTION
**Pull request by mrutkowsky**
Please download modules in order to use this feature:
xlsxwriter
io

I've also added .gitkeep files in empty directories instead of .txt and changed the way of reading files in those folders for not displaying useless files for users

### CODE REVIEW

File: app/narwos.py [hunk 1]
```diff
@@ -112,10 +114,12 @@ def index():
 
     validated_files = os.listdir(
         PATH_TO_VALID_FILES)
+        
+    validated_files_to_show = [file for file in validated_files if file != '.gitkeep']
```
Comment (MatPatCarry) [inline app/narwos.py:118]: If you decide to include such logic, you have to remember not only exclude these files from GUI, but also block to process them on API, but of course idea is ok

File: app/narwos.py [hunk 2]
```diff
@@ -112,10 +114,12 @@ def index():
...
 
     return render_template(
         "index.html", 
-        files=validated_files,
```
Comment (MatPatCarry) [inline app/narwos.py:118]: definetely, you do not want to pass two lists of files if one of them you do not show and do not need -> pass one list with excluded .gitkeep files

File: app/narwos.py [hunk 3]
```diff
@@ -273,6 +277,52 @@ def show_clusters():
...
+    show_filter = request.form.get('show_filter')
+    if show_filter:
+        return redirect(url_for('show_filter', show_filter=show_filter))
+    else:
+        return redirect(url_for("index", message=f"Cannot show the filtering!"))
+
+@app.route('/show_filters', methods=['GET'])
+def show_filter():
+
+    df = pd.read_csv(PATH_TO_CURRENT_DF)
```
Comment (MatPatCarry) [inline app/narwos.py:292]: Finally this file will be .gzip.pqrquet in order to provide quick reading and saving

Comment (MatPatCarry) [reply app/narwos.py:292]: And do not use function from pandas, check our utils and search for read_file()

File: app/narwos.py [hunk 4]
```diff
@@ -273,6 +277,52 @@ def show_clusters():
...
+    if isinstance(df, pd.DataFrame):
+
+        return render_template('filtering.html', columns=df.columns)
+    
+    return 'Nothing to show here'
+
+@app.route('/filter', methods=['POST'])
+def filter_data():
+    filters = request.get_json()
+    filtered_df = pd.read_csv(PATH_TO_CURRENT_DF)
```
Comment (MatPatCarry) [inline app/narwos.py:303]: the same as above and why do you have to read df twice? Is it not possible to keep it memory? But do not use flask.session object becuase object is to big, but you can try to experiment with g object and serialization, maybe you can find a better aproach then reading file twice

File: app/narwos.py [hunk 5]
```diff
@@ -273,6 +277,52 @@ def show_clusters():
...
+    filtered_df = filtered_df.astype(str)
+    logger.info(
+        f"""Columns of df to filtr:{filtered_df.columns}""")
+    for filter_data in filters:
+        column = filter_data['column']
```
Comment (MatPatCarry) [inline app/narwos.py:308]: If you want to retrieve value from dict, use .get(), when you try to retrieve value for key which does not exists, you would get error, you have to service this, maybe add some logic with try, except if these filters are not found.

File: app/narwos.py [hunk 6]
```diff
@@ -273,6 +277,52 @@ def show_clusters():
...
+        value = filter_data['value']
+        filtered_df = filtered_df[filtered_df[column] == value]
```
Comment (MatPatCarry) [inline app/narwos.py:310]: get filters first, most importantly use .LOC - the fastest or even think of situation where there is more then 20k rows - here you can use eval() - read about it. Do not apply filters in loop - combine them into one big condition and apply one big filter once - maybe read about pd.query here

File: app/narwos.py [hunk 7]
```diff
@@ -273,6 +277,52 @@ def show_clusters():
...
+
+    # Prepare the CSV file for download
+    output = io.BytesIO()
```
Comment (MatPatCarry) [inline app/narwos.py:313]: not include such logic here - put them in utils, you can try to write a func which returns for user raport in excel or csv - it depends on user choice.

File: app/narwos.py [hunk 8]
```diff
@@ -273,6 +277,52 @@ def show_clusters():
...
+    with pd.ExcelWriter(output, engine='xlsxwriter') as writer:
+        filtered_df.to_excel(writer, index=False, sheet_name='Filtered Data')
+        writer.close()
+    output.seek(0)
+
+    return send_file(
```
Comment (MatPatCarry) [inline app/narwos.py:319]: ok, but hardcoded, download_name definetely in CONFIG.yaml, try to catch situations which I have written above.

File: app/templates/filtering.html [hunk 9]
```diff
@@ -0,0 +1,129 @@
+<!DOCTYPE html>
```
Comment (MatPatCarry) [inline app/templates/filtering.html:1]: I do not have knowledge to review html, javascript, but it looks ok

**Review by MatPatCarry — CHANGES_REQUESTED** [review]
Please include my comments, try to implement it then we can merge ;) Overall really good job Maciej

**Comment by MatPatCarry** [comment]
image
little change needed

**Comment by MatPatCarry** [comment]
image
x, y are not for filtering, they can be excluded and very important issue -> Questioned date should be some calendar, slide bar or sth like that in the future

**Comment by MatPatCarry** [comment]
For categories, for example OS it should be a good idea to show which values are available

**Comment by MatPatCarry** [comment]
after applying filter user does not n=know if filter was applied successfully

File: app/narwos.py (hunk 3 above)
Comment (mrutkowsky) [reply app/narwos.py:292]: Read file do not work for this file extension idk why. Please check the function for returning data for clustering vizualization- Logic looks the same

**Pull request merged**